src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java

```diff
@@ -0,0 +1,215 @@
+                                       .build());
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    recordTransformer = new StructuredToAvroTransformer(config.schema, config.fieldNames);
+  }
+
+  @Override
+  protected void addPartitionedFileSetProperties(PartitionedFileSetProperties.Builder properties) {
+    FileSetUtil.configureAvroFileSet(config.schema, properties);
+    properties.addAll(FileSetUtil.getAvroCompressionConfiguration(config.compressionCodec, config.schema, true));
+  }
+
+  @Override
+  public void transform(StructuredRecord input,
+                        Emitter<KeyValue<AvroKey<GenericRecord>, NullWritable>> emitter) throws Exception {
+    emitter.emit(new KeyValue<>(new AvroKey<>(recordTransformer.transform(input)), NullWritable.get()));
+  }
+
+  /**
+   * Config for DynamicPartitionFileSetAvroSink
+   */
+  public static class DynamicPartitionAvroSinkConfig extends PartitionedFileSetSinkConfig {
+    public static final String FIELD_NAMES_PROPERTY_KEY = "fieldNames";
+
+    @Description(SCHEMA_DESC)
+    private String schema;
+    @Description(FIELD_DESC)
+    private String fieldNames;
+
+    public DynamicPartitionAvroSinkConfig(String name, String schema, String fieldNames,
+                                             @Nullable String basePath, @Nullable String compressionCodec) {
+      super(name, basePath, compressionCodec);
+      this.schema = schema;
+      this.fieldNames = fieldNames;
+    }
+  }
+
+  /**
+   * Dynamic partitioner that creates partitions based on a list of fields in each record.
+   */
+  public static final class FieldValueDynamicPartitioner extends DynamicPartitioner<AvroKey<GenericRecord>, NullWritable> {
+    private String[] fieldNames;
+
+    @Override
+    public void initialize(MapReduceTaskContext mapReduceTaskContext) {
+      if (mapReduceTaskContext.getPluginProperties(DynamicPartitionFileSetAvroSink.NAME) == null) {
+        throw new IllegalArgumentException("Could not find a plugin with the name: " +
+                                             DynamicPartitionFileSetAvroSink.NAME + " in the list of plugins.");
+      }
+      // Need a better way to do this. [CDAP-7058]
+      fieldNames = mapReduceTaskContext
+        .getPluginProperties(DynamicPartitionFileSetAvroSink.NAME)
+        .getProperties().get(DynamicPartitionAvroSinkConfig.FIELD_NAMES_PROPERTY_KEY).split(",");
+    }
+
+    @Override
+    public PartitionKey getPartitionKey(AvroKey<GenericRecord> key, NullWritable value) {
+      PartitionKey.Builder keyBuilder = PartitionKey.builder();
+      for (int i = 0; i < fieldNames.length; i++) {
+        if (key.datum().get(fieldNames[i]) == null) {
+          // This call will result in an exception but there's nothing else I can do. [CDAP-7053]
+          keyBuilder.addStringField(fieldNames[i], null);
```

Comment:
to avoid this, maybe there should be a check to make sure the partition columns are marked as "not nullable"